src/util/git.js

```diff
@@ -385,89 +367,82 @@ export default class Git {
-      this.supportsArchive = false;
+  async useDefaultBranchWithSymref(): Promise<void> {
+    const stdout = await Git.spawn(['ls-remote', '--symref', this.gitUrl.repository, 'HEAD']);
+    const lines = stdout.split('\n');
+    const [, ref] = lines[0].split(/\s+/);
+    const [sha] = lines[1].split(/\s+/);
+    this.ref = ref;
+    this.hash = sha;
+  }
 
-      if (!this.fetched) {
-        // in fact, `git archive` can't be used, and we haven't fetched the project yet. Do it now.
-        await this.fetch();
-      }
-      return (this.ref = this.hash = hash);
+  async useDefaultBranchWithLegacyGit(refs: GitRefs): Promise<void> {
+    const stdout = await Git.spawn(['ls-remote', this.gitUrl.repository, 'HEAD']);
+    const [sha] = stdout.split(/\s+/);
+    const resolvedResult = await resolveVersion(this.config, sha, refs);
+    if (resolvedResult !== false && !resolvedResult.defaultBranch) {
+      this.hash = resolvedResult.sha;
+      this.ref = resolvedResult.ref || '';
+    } else {
+      this.hash = sha;
+      this.ref = '';
     }
+  }
 
-    const ref = await this.findResolution(hash, names);
-    const commit = refs[ref];
-    if (commit) {
-      this.ref = ref;
-      return (this.hash = commit);
-    } else {
-      throw new MessageError(this.reporter.lang('couldntFindMatch', ref, names.join(','), this.gitUrl.repository));
+  async useDefaultBranch(refs: GitRefs): Promise<void> {
+    try {
+      await this.useDefaultBranchWithSymref();
+    } catch (err) {
+      // older versions of git don't support "--symref"
+      await this.useDefaultBranchWithLegacyGit(refs);
     }
   }
 
   /**
-   * Parse Git ref lines into hash of tag names to SHA hashes
+   * TODO description
    */
 
-  static parseRefs(stdout: string): GitRefs {
-    // store references
-    const refs = {};
-
-    // line delimited
-    const refLines = stdout.split('\n');
-
-    for (const line of refLines) {
-      const match = gitRefLineRegex.exec(line);
-
-      if (match) {
-        const [, sha, tagName] = match;
-
-        // As documented in gitrevisions:
-        //   https://www.kernel.org/pub/software/scm/git/docs/gitrevisions.html#_specifying_revisions
-        // "A suffix ^ followed by an empty brace pair means the object could be a tag,
-        //   and dereference the tag recursively until a non-tag object is found."
-        // In other words, the hash without ^{} is the hash of the tag,
-        //   and the hash with ^{} is the hash of the commit at which the tag was made.
-        const name = removeSuffix(tagName, '^{}');
+  async setRef(refs: GitRefs): Promise<string> {
+    // get commit ref
+    const {hash: version} = this;
 
-        refs[name] = sha;
-      }
+    const resolvedResult = await resolveVersion(this.config, version, refs);
+    if (resolvedResult === false) {
+      throw new MessageError(
+        this.reporter.lang('couldntFindMatch', version, Object.keys(refs).join(','), this.gitUrl.repository),
```

Comment:
The `refs/merge-requests/*` had nothing to do here, it should be solved with latest commit.